torch/csrc/Generator.cpp

```diff
@@ -116,37 +116,60 @@ static PyObject* THPGenerator_setState(PyObject* _self, PyObject* _new_state) {
   END_HANDLE_TH_ERRORS
 }
 
-static PyObject* THPGenerator_manualSeed(PyObject* _self, PyObject* seed) {
-  HANDLE_TH_ERRORS
-  auto self = (THPGenerator*)_self;
-  auto generator = self->cdata;
-  THPUtils_assert(
-      THPUtils_checkLong(seed),
-      "manual_seed expected a long, "
-      "but got %s",
-      THPUtils_typename(seed));
-  // See Note [Acquire lock when using random generators]
-  std::lock_guard<std::mutex> lock(generator.mutex());
+uint64_t unpack_int_pyobj(PyObject* obj) {
   // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
-  uint64_t seed_unpacked;
+  uint64_t obj_unpacked;
   try {
     // First try to interpret as unsigned long
-    seed_unpacked = THPUtils_unpackUInt64(seed);
+    obj_unpacked = THPUtils_unpackUInt64(obj);
   } catch (...) {
     if (PyErr_ExceptionMatches(PyExc_OverflowError)) {
-      // If an overflow happened, then the seed could be negative,
+      // If an overflow happened, then the obj could be negative,
       // so try to interpret it as signed long
       PyErr_Clear();
-      int64_t seed_unpacked_signed = THPUtils_unpackLong(seed);
-      seed_unpacked = *(reinterpret_cast<uint64_t*>(&seed_unpacked_signed));
+      int64_t obj_unpacked_signed = THPUtils_unpackLong(obj);
+      obj_unpacked = *(reinterpret_cast<uint64_t*>(&obj_unpacked_signed));
     } else {
       // If any other type of exception happened, rethrow it
       throw;
     }
   }
+  return obj_unpacked;
+}
+
+static PyObject* THPGenerator_manualSeed(PyObject* _self, PyObject* seed) {
+  HANDLE_TH_ERRORS
+  auto self = (THPGenerator*)_self;
+  auto generator = self->cdata;
+  THPUtils_assert(
+      THPUtils_checkLong(seed),
+      "manual_seed expected a long, "
+      "but got %s",
+      THPUtils_typename(seed));
+  uint64_t seed_unpacked = unpack_int_pyobj(seed);
+  // See Note [Acquire lock when using random generators]
```

Comment:
I wonder if you can do thread-safe generators by making thread affine/thread local copies of underlying data. cc @albanD